### PR TITLE
Rewrite vm service layer in flutter_tools

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] some checked mode fixes
- [x] Minimal port of the VMService layer from Observatory
- [x] Support VM, Isolate, FlutterView, ServiceEvent service objects explicitly
- [x] ServiceMap catch all for any other objects
- [x] ResidentRunner keeps track of the current FlutterView and the ui isolate inside of it
- [x] Move a bunch of operations off of the VMService class and onto the VM or Isolate classes

### CODE REVIEW

**Comment by johnmccutchan** [comment]
@devoncarew @turnidge

File: packages/flutter_tools/lib/src/commands/trace.dart [hunk 1]
```diff
@@ -95,40 +95,41 @@ class TraceCommand extends FlutterCommand {
...
+      vmService.onTimelineEvent.listen((ServiceEvent timelineEvent) {
+        List<Map<String, dynamic>> events = timelineEvent.timelineEvents;
         for (Map<String, dynamic> event in events) {
           if (event['name'] == kFirstUsefulFrameEventName)
             whenFirstFrameRendered.complete();
         }
       });
-      await observatory.streamListen('Timeline');
+      // TODO(johnmccutchan): Is this necessary?
+      // await observatory.streamListen('Timeline');
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/trace.dart]: This was necessary before; now `onTimelineEvent` will automatically call a streamListen. I think this is no longer necessary -

File: packages/flutter_tools/lib/src/hot.dart [hunk 2]
```diff
@@ -477,16 +475,16 @@ class HotRunner extends ResidentRunner {
       } else {
         flutterUsage.sendEvent('hot', 'reload');
       }
-    } catch (errorMessage) {
+    } catch (errorMessage, st) {
       reloadStatus.stop(showElapsedTime: true);
-      printError('Hot reload failed:\n$errorMessage');
+      printError('Hot reload failed:\n$errorMessage\n$st');
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/hot.dart]: You can do a `new Chain.forTrace(st).terse` here to get a more readable stack trace.

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 3]
```diff
@@ -40,155 +32,543 @@ class VMService {
...
+    _getEventController(streamId).add(event);
   }
 
-  Future<Null> populateIsolateInfo() async {
-    // Calling this has the side effect of populating the isolate information.
-    await waitFirstIsolate;
+  Future<Null> _streamListen(String streamId) async {
+    if (!_listeningFor.contains(streamId)) {
+      _listeningFor.add(streamId);
+      peer.sendRequest('streamListen',
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/vmservice.dart]: We could await this, or change _streamListen to return void.

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 4]
```diff
@@ -40,155 +32,543 @@ class VMService {
...
-        _addIsolate(isolate);
-    });
+bool _isServiceMap(Map<String, dynamic> m) {
+  return (m != null) && (m['type'] != null);
+}
+bool _hasRef(String type) => (type != null) && type.startsWith('@');
+String _stripRef(String type) => (_hasRef(type) ? type.substring(1) : type);
 
-    return _waitFirstIsolateCompleter.future;
+void _upgradeCollection(dynamic collection,
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/vmservice.dart]: Perhaps some docs or a description of what an upgrade means? Plain collections to wrapped service objects?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 5]
```diff
@@ -40,155 +32,543 @@ class VMService {
...
+  Future<Isolate> getIsolate(String isolateId) {
+    if (!loaded) {
+      // Trigger a VM load, then get the isolate. Ignore any errors.
+      return load().then((_) => getIsolate(isolateId)).catchError((_) => null);
+    }
+    return new Future<Isolate>.value(_isolateCache[isolateId]);
+  }
+
+  /// Invoke the RPC and return the raw response.
+  Future<Map<String, dynamic>> invokeRpcRaw(
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/vmservice.dart]: I like having a wrapper for responses - `Response` as opposed to `Map<String, dynamic>` - it makes method signatures simpler in a lot of places, and you can hang semantic methods off the object; but YMMV.

**Comment by devoncarew** [comment]
lgtm

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 4 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/vmservice.dart]: This is the magic function which recursively walks the raw response from the service protocol and upgrades values in the map into full service objects. It's an implementation detail but I'll add some documentation here.

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 5 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/vmservice.dart]: It's expected that most people will call `invokeRpc` and not `invokeRpcRaw`. `invokeRpc` always returns a ServiceObject (with ServiceMap as the fallback). There are too many callers of `invokeRpcRaw` right now but that will go down over time as I do more cleanups.

File: packages/flutter_tools/lib/src/commands/trace.dart (hunk 1 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/commands/trace.dart]: Removed.

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 3 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/vmservice.dart]: awaited

**Pull request merged**